File: scripts/generate-pwa-icons.mjs
import fs from "node:fs";
import path from "node:path";
import { createCanvas } from "canvas";

// Ensure public/icons directory exists
const iconsDir = path.join(process.cwd(), "public", "icons");
if (!fs.existsSync(iconsDir)) {
  fs.mkdirSync(iconsDir, { recursive: true });
}

/**
 * Creates a CallDesk PWA Icon with a dark theme gradient background (#09090b to #18181b),
 * an emerald glow (#10b981), and a phone/dial icon graphic.
 */
function generateIcon(size, isMaskable = false, isApple = false) {
  const canvas = createCanvas(size, size);
  const ctx = canvas.getContext("2d");

  // Background
  const gradient = ctx.createLinearGradient(0, 0, size, size);
  gradient.addColorStop(0, "#09090b");
  gradient.addColorStop(1, "#18181b");
  ctx.fillStyle = gradient;

  if (isApple) {
    // Apple Touch Icon: Filled square (iOS applies rounded corners automatically)
    ctx.fillRect(0, 0, size, size);
  } else if (!isMaskable) {
    // Standard icon: Rounded rectangle
    const cornerRadius = size * 0.22;
    ctx.beginPath();
    ctx.roundRect(0, 0, size, size, cornerRadius);
    ctx.fill();
  } else {
    // Maskable icon: Full canvas fill for Android adaptive icon safe zone
    ctx.fillRect(0, 0, size, size);
  }

  // Safe area padding for maskable icons (inner 80%)
  const scale = isMaskable ? 0.75 : 0.85;
  const centerX = size / 2;
  const centerY = size / 2;

  // Outer Emerald Glow Ring
  ctx.save();
  ctx.beginPath();
  ctx.arc(centerX, centerY, size * 0.32 * scale, 0, Math.PI * 2);
  ctx.strokeStyle = "#059669";
  ctx.lineWidth = size * 0.04 * scale;
  ctx.stroke();
  ctx.restore();

  // Inner Emerald Circle
  ctx.save();
  ctx.beginPath();
  ctx.arc(centerX, centerY, size * 0.26 * scale, 0, Math.PI * 2);
  ctx.fillStyle = "#10b981";
  ctx.fill();
  ctx.restore();

  // Phone Handle Symbol (White)
  ctx.save();
  ctx.fillStyle = "#09090b";
  ctx.translate(centerX, centerY);

  // Phone receiver shape
  const iconScale = size * 0.003 * scale;
  ctx.beginPath();
  ctx.arc(-8 * iconScale, -10 * iconScale, 6 * iconScale, 0, Math.PI * 2);
  ctx.arc(10 * iconScale, 8 * iconScale, 6 * iconScale, 0, Math.PI * 2);
  ctx.fill();

  ctx.beginPath();
  ctx.moveTo(-10 * iconScale, -6 * iconScale);
  ctx.quadraticCurveTo(0, -12 * iconScale, 6 * iconScale, -10 * iconScale);
  ctx.lineTo(10 * iconScale, 6 * iconScale);
  ctx.quadraticCurveTo(12 * iconScale, 0, -6 * iconScale, 10 * iconScale);
  ctx.closePath();
  ctx.fillStyle = "#ffffff";
  ctx.fill();

  ctx.restore();

  return canvas.toBuffer("image/png");
}

console.log("Generating PWA Icons for CallDesk...");

try {
  fs.writeFileSync(path.join(iconsDir, "icon-192x192.png"), generateIcon(192));
  fs.writeFileSync(path.join(iconsDir, "icon-512x512.png"), generateIcon(512));
  fs.writeFileSync(path.join(iconsDir, "icon-maskable-192x192.png"), generateIcon(192, true));
  fs.writeFileSync(path.join(iconsDir, "icon-maskable-512x512.png"), generateIcon(512, true));
  fs.writeFileSync(path.join(iconsDir, "apple-touch-icon.png"), generateIcon(180, false, true));
  console.log("[SUCCESS] Generated 5 PWA icons in public/icons/");
} catch (err) {
  console.error("Canvas icon generation error:", err.message);
  // Fallback: Write SVG-based static PNG or check canvas availability
}

File: scripts/generate_pwa_icons.py
import os
from PIL import Image, ImageDraw

icons_dir = os.path.join(os.getcwd(), "public", "icons")
os.makedirs(icons_dir, exist_ok=True)

def create_pwa_icon(size, is_maskable=False, is_apple=False):
    # Base Canvas
    img = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)
    
    # Fill background
    bg_color = (9, 9, 11, 255) # #09090b
    
    if is_apple or is_maskable:
        # Full square for Apple and Android Maskable adaptive background
        draw.rectangle([0, 0, size, size], fill=bg_color)
    else:
        # Standard rounded rectangle
        corner_radius = int(size * 0.22)
        draw.rounded_rectangle([0, 0, size, size], radius=corner_radius, fill=bg_color)
        
    scale = 0.75 if is_maskable else 0.85
    center_x = size // 2
    center_y = size // 2
    
    # Outer emerald ring
    ring_radius = int(size * 0.32 * scale)
    ring_width = int(size * 0.04 * scale)
    draw.ellipse(
        [center_x - ring_radius, center_y - ring_radius, center_x + ring_radius, center_y + ring_radius],
        outline=(5, 150, 105, 255),
        width=ring_width
    )
    
    # Inner emerald circle
    inner_radius = int(size * 0.25 * scale)
    draw.ellipse(
        [center_x - inner_radius, center_y - inner_radius, center_x + inner_radius, center_y + inner_radius],
        fill=(16, 185, 129, 255) # #10b981
    )
    
    # Phone Handle Icon in center
    phone_radius = int(inner_radius * 0.45)
    
    # Left receiver circle
    draw.ellipse(
        [center_x - phone_radius - 5, center_y - 10, center_x - phone_radius + 15, center_y + 10],
        fill=(9, 9, 11, 255)
    )
    # Right receiver circle
    draw.ellipse(
        [center_x + phone_radius - 15, center_y - 10, center_x + phone_radius + 5, center_y + 10],
        fill=(9, 9, 11, 255)
    )
    # Connecting handle bar
    draw.line(
        [center_x - phone_radius + 5, center_y, center_x + phone_radius - 5, center_y],
        fill=(9, 9, 11, 255),
        width=int(size * 0.05 * scale)
    )
    
    # Inner white accent ring inside phone badge
    draw.ellipse(
        [center_x - int(inner_radius * 0.3), center_y - int(inner_radius * 0.3), center_x + int(inner_radius * 0.3), center_y + int(inner_radius * 0.3)],
        outline=(255, 255, 255, 255),
        width=int(size * 0.02 * scale)
    )

    return img

print("Generating 5 crisp PWA PNG icons...")
create_pwa_icon(192).save(os.path.join(icons_dir, "icon-192x192.png"))
create_pwa_icon(512).save(os.path.join(icons_dir, "icon-512x512.png"))
create_pwa_icon(192, is_maskable=True).save(os.path.join(icons_dir, "icon-maskable-192x192.png"))
create_pwa_icon(512, is_maskable=True).save(os.path.join(icons_dir, "icon-maskable-512x512.png"))
create_pwa_icon(180, is_apple=True).save(os.path.join(icons_dir, "apple-touch-icon.png"))
print("[SUCCESS] All 5 PWA icons generated in public/icons/")
